docs/man/memcached_stats.3
.TH "MEMCACHED_STATS" "3" "September 16, 2012" "1.0.10" "libmemcached"
.SH NAME
memcached_stats \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
Get memcached statistics
.SH SYNOPSIS
.sp
#include <libmemcached/memcached.h>
.INDENT 0.0
.TP
.B memcached_stat_st
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t (*memcached_stat_fn)(memcached_server_instance_st\fI\ server\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, const char\fI\ *value\fP, size_t\fI\ value_length\fP, void\fI\ *context\fP)
.UNINDENT
.INDENT 0.0
.TP
.B \fI\%memcached_stat_st\fP *memcached_stat(memcached_st\fI\ *ptr\fP, char\fI\ *args\fP, memcached_return_t\fI\ *error\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_stat_servername(\fI\%memcached_stat_st\fP\fI\ *stat\fP, char\fI\ *args\fP, const char\fI\ *hostname\fP, in_port_t\fI\ port\fP)
.UNINDENT
.INDENT 0.0
.TP
.B char * memcached_stat_get_value(memcached_st\fI\ *ptr\fP, \fI\%memcached_stat_st\fP\fI\ *stat\fP, const char\fI\ *key\fP, memcached_return_t\fI\ *error\fP)
.UNINDENT
.INDENT 0.0
.TP
.B char ** memcached_stat_get_keys(memcached_st\fI\ *ptr\fP, \fI\%memcached_stat_st\fP\fI\ *stat\fP, memcached_return_t\fI\ *error\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_stat_execute(memcached_st\fI\ *memc\fP, const char\fI\ *args\fP, \fI\%memcached_stat_fn\fP\fI\ func\fP, void\fI\ *context\fP)
.UNINDENT
.sp
Compile and link with \-lmemcached
.SH DESCRIPTION
.sp
libmemcached(3) has the ability to query a memcached server (or collection
of servers) for their current state. Queries to find state return a
\fI\%memcached_stat_st\fP structure. You are responsible for freeing this structure. While it is possible to access the structure directly it is not advisable. \fI\%memcached_stat_get_value()\fP has been provided to query the structure.
.sp
\fI\%memcached_stat_execute()\fP uses the servers found in \fI\%memcached_stat_st\fP and executes a "stat" command on each server. args is an optional argument that can be passed in to modify the behavior of "stats". You will need to supply a callback function that will be supplied each pair of values returned by
the memcached server.
.sp
\fI\%memcached_stat()\fP fetches an array of \fI\%memcached_stat_st\fP structures containing the state of all available memcached servers. The return value must be freed by the calling application. If called with the \fBMEMCACHED_BEHAVIOR_USE_UDP\fP behavior set, a NULL value is returned and the error parameter is set to \fBMEMCACHED_NOT_SUPPORTED\fP.
.sp
\fI\%memcached_stat_servername()\fP can be used standalone without a \fBmemcached_st\fP to obtain the state of a particular server.  "args" is used to define a particular state object (a list of these are not provided for by either
the \fI\%memcached_stat_get_keys()\fP call nor are they defined in the memcached protocol). You must specify the hostname and port of the server you want to
obtain information on.
.sp
\fI\%memcached_stat_get_value()\fP returns the value of a particular state key. You specify the key you wish to obtain.  The key must be null terminated.
.sp
\fI\%memcached_stat_get_keys()\fP returns a list of keys that the server has state objects on. You are responsible for freeing this list.
.sp
A command line tool, memstat(1), is provided so that you do not have to write
an application to do this.
.SH RETURN
.sp
Varies, see particular functions.
.sp
Any method returning a \fI\%memcached_stat_st\fP expects you to free the
memory allocated for it.
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH AUTHOR
.sp
Brian Aker, <\fI\%brian@tangent.org\fP>
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP \fImemcached_strerror(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2012, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
1.0.11 Sun Sep 16 20:32:13 EDT 2012
* Removed custom version of memcached.
* Updated hardening rules.
* Fixed a case where the return error from a socket connection differred from that of a TCP/IP socket.

1.0.10 Sun Jul 29 21:50:15 PDT 2012
* --disable-assert has been removed from configure, and --enable-assert has been added in its place.
* Compiling fixes for Clang on OSX Mountain Lion.

1.0.9 Wed Jul  4 22:46:19 EDT 2012
* Faster close on socket.
* Instance allocation is now seperated from server interface. This should allow for a better preservation of ABI compliance from now on.
* Fix close on exec bug.
* Numerous other bug fixes.

1.0.8 Tue May 22 15:06:04 EDT 2012
* Added support for setting options via ENV variable LIBMEMCACHED
* Fix corner case on last used result.

1.0.7 Sat Apr 28 00:48:29 PDT 2012
* Add API call for exist calls.
* Update all license files to be BSD.

1.0.6 Sat Apr  7 18:26:49 PDT 2012
* Fixes for gcc 4.7, lp:961812
* Fix for restart issue that happens under testing.
* Fix for lp:962815.
* Support for transparent AES encryption.

1.0.5 Tue Mar 13 22:56:47 PDT 2012
* Fixes for OSX.
* Version is now parsed directly in the parser, which makes buffered operations now work with it..
* memstat has been extended so that it can be used to find the version of the server.
* Update documentation.
* Fixes for compile issues on Debian and Ubuntu

1.0.4 Thu Jan 26 22:33:54 PST 2012
* Fix for memcached_dump().
* Additional testing for memcached_stat_execute().

1.0.3 Sun Jan  8 18:11:36 PST 2012
* Increased size of sort buffer used during Ketama.
* Added support for new behavior to handle dead servers.
* Overall haul of UDP IO.
* Fixed C compile issue with memcached_exist()
* Numerous bug fixes.
* Clang support for OSX.
* All commands now using vector send support.

1.0.2 Mon Oct 24 08:00:48 PDT 2011
* Dropped libmemcached/memcached_util.h (undocumented header file)
* Added memcached_touch() and memcached_touch_by_key()
* UDP support restructured to toggle on a complete memcached_st structure.

0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet

0.52 Sun Sep 11 22:16:08 PDT 2011
  * Build fixes for Ubuntu/Suse.
  * Fixes for OSX Lion.
  * Bug fix for looping back through dns lookups under certain failures.
  * Fixes related to dead server failures.

0.51 Thu Jul 21 11:08:58 PDT 2011
  * memcached_callback_set() now takes its data argument as const
  * Update to tests.
  * Fix in parser for port number.

0.50 Mon Jun 20 10:36:57 PDT 2011
 * Updates to C++ interface 
 * Custom free allocators need to now check for value before calling free.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).
 * Fix for stats structure.
 * Updates to documentation.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).

0.49 Thu Apr 14 08:43:37 PDT 2011
  * Fix calls to auto methods so that if value is not passed in nothing bad happens.
  * New parser calls for generating memcached_st objects.
  * New error system.
  * New flow control for messages means faster get/set calls.
  * Added new documentation system.
  * A behavior change has been now made that if you specify a weight for any server, we enable the weight flag and do weight balancing.  
  * Added MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS to simplify the setting of AUTO REJECT for servers.

0.48 Tue Mar 15 23:05:18 PDT 2011
  * Fix memory leak in server parse.
  * Move test framework out to be its own library (easier to work with Gearman).


0.47 Wed Feb 23 18:46:48 PST 2011
  * Additional fixes for OpenBSD.
  * Bug fix 677609, 456080.
  * SIGPIPE fix for Linux send(). 
  * memcapable can now test ascii or binary based on flags.
  * Additional build fixes for SASL.


0.46 Mon Feb 14 10:28:01 PST 2011
  * Fixes a number of corner case bugs.
  * Fixes related to OpenBSD.
  * Better testing for protocol version.
  * Removes special case infinite wait on blocking setup.

0.45 Tue Feb  8 16:02:06 PST 2011
  * Add support for systemtap

0.44 Wed Sep 22 21:57:57 PDT 2010
  * Windows bug fixes.
  * Hudson port support in test harness.
  * Improved portability of test hanrness.
  * SASL fixes.

0.43 Wed Jul 28 16:29:47 PDT 2010
  * Added --args to memstat so that a greater range of values can be returned.
  * Prelimanary support for Windows.
  * memcached_stat_execute() merged. 

0.42 Tue Jul  6 12:29:50 PDT 2010
  * Mistake in libtool caused issue with library version

0.41 Wed Jun 30 13:15:19 PDT 2010
  * Added --file for memcat.
  * Added limemcached_ping() to libmemcached_util
  * Bugfix for some cases where connect would have issues with timeout.
  * Wrong value for errno given as error on an IO failure inside of poll.
  * Bug fix for issue where multiple interfaces with bad DNS were not being caught.

0.40 Thu Apr 22 19:01:25 PDT 2010
  * Placed retry logic in for busted resolvers
  * Add an ignore for SIGPIPE to solve OSX issues.
  * A couple of fixed for memcached_light server.
  * Updated to debug mode to track io_wait

0.39 Tue Apr  6 12:35:13 PDT 2010
  * Add support for prefix keys to binary protocol.
  * Remove the undocumented call memcached_server_remove().
  * The undocumented call memcached_server_by_key() now returns const.
  * memcached_server_error_reset() has been deprecated.
  * memcached_server_list() has been deprecated. Use memcached_server_cursor() to walk the servers found in a memcached_st() structure.
  * memcached_verbosity() can now be run concurrently with other operations.
  * SASL support.
  * Fixes memory leak found in EJECT HOSTS.

0.38 Wed Feb 10 09:40:33 PST 2010
  * C++ interface for libhashkit.
  * Modified memcached_set_memory_allocators() so that it requires a context pointer.
  * memcached_clone() now runs 5 times faster.
  * Functions used for callbacks are now given const memcached_st.

libmemcached/memcached_stats.c
/*
*/

#include "common.h"

static const char *memcached_stat_keys[] = {
  "pid",
  "uptime",
  "time",
  "version",
  "pointer_size",
  "rusage_user",
  "rusage_system",
  "curr_items",
  "total_items",
  "bytes",
  "curr_connections",
  "total_connections",
  "connection_structures",
  "cmd_get",
  "cmd_set",
  "get_hits",
  "get_misses",
  "evictions",
  "bytes_read",
  "bytes_written",
  "limit_maxbytes",
  "threads",
  NULL
};


static memcached_return_t set_data(memcached_stat_st *memc_stat, char *key, char *value)
{

  if (strlen(key) < 1) 
  {
    WATCHPOINT_STRING(key);
    return MEMCACHED_UNKNOWN_STAT_KEY;
  }
  else if (!strcmp("pid", key))
  {
    memc_stat->pid= (uint32_t) strtol(value, (char **)NULL, 10);
  }
  else if (!strcmp("uptime", key))
  {
    memc_stat->uptime= (uint32_t) strtol(value, (char **)NULL, 10);
  }
  else if (!strcmp("time", key))
  {
    memc_stat->time= (uint32_t) strtol(value, (char **)NULL, 10);
  }
  else if (!strcmp("version", key))
  {
    memcpy(memc_stat->version, value, strlen(value));
    memc_stat->version[strlen(value)]= 0;
  }
  else if (!strcmp("pointer_size", key))
  {
    memc_stat->pointer_size= (uint32_t) strtol(value, (char **)NULL, 10);
  }
  else if (!strcmp("rusage_user", key))
  {
    char *walk_ptr;
    for (walk_ptr= value; (!ispunct(*walk_ptr)); walk_ptr++);
    *walk_ptr= 0;
    walk_ptr++;
    memc_stat->rusage_user_seconds= (uint32_t) strtol(value, (char **)NULL, 10);
    memc_stat->rusage_user_microseconds= (uint32_t) strtol(walk_ptr, (char **)NULL, 10);
  }
  else if (!strcmp("rusage_system", key))
  {
    char *walk_ptr;
    for (walk_ptr= value; (!ispunct(*walk_ptr)); walk_ptr++);
    *walk_ptr= 0;
    walk_ptr++;
    memc_stat->rusage_system_seconds= (uint32_t) strtol(value, (char **)NULL, 10);
    memc_stat->rusage_system_microseconds= (uint32_t) strtol(walk_ptr, (char **)NULL, 10);
  }
  else if (!strcmp("curr_items", key))
  {
    memc_stat->curr_items= (uint32_t) strtol(value, (char **)NULL, 10); 
  }
  else if (!strcmp("total_items", key))
  {
    memc_stat->total_items= (uint32_t) strtol(value, (char **)NULL, 10);
  }
  else if (!strcmp("bytes_read", key))
  {
    memc_stat->bytes_read= (uint32_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("bytes_written", key))
  {
    memc_stat->bytes_written= (uint32_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("bytes", key))
  {
    memc_stat->bytes= (uint32_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("curr_connections", key))
  {
    memc_stat->curr_connections= (uint32_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("total_connections", key))
  {
    memc_stat->total_connections= (uint32_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("connection_structures", key))
  {
    memc_stat->connection_structures= (uint32_t) strtol(value, (char **)NULL, 10);
  }
  else if (!strcmp("cmd_get", key))
  {
    memc_stat->cmd_get= (uint64_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("cmd_set", key))
  {
    memc_stat->cmd_set= (uint64_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("get_hits", key))
  {
    memc_stat->get_hits= (uint64_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("get_misses", key))
  {
    memc_stat->get_misses= (uint64_t)strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("evictions", key))
  {
    memc_stat->evictions= (uint64_t)strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("limit_maxbytes", key))
  {
    memc_stat->limit_maxbytes= (uint64_t) strtoll(value, (char **)NULL, 10);
  }
  else if (!strcmp("threads", key))
  {
    memc_stat->threads= (uint32_t) strtol(value, (char **)NULL, 10);
  }
  else if (!(strcmp("delete_misses", key) == 0 ||/* New stats in the 1.3 beta */
             strcmp("delete_hits", key) == 0 ||/* Just swallow them for now.. */
             strcmp("incr_misses", key) == 0 ||
             strcmp("incr_hits", key) == 0 ||
             strcmp("decr_misses", key) == 0 ||
             strcmp("decr_hits", key) == 0 ||
             strcmp("cas_misses", key) == 0 ||
             strcmp("cas_hits", key) == 0 ||
             strcmp("cas_badval", key) == 0 ||
             strcmp("cmd_flush", key) == 0 ||
             strcmp("accepting_conns", key) == 0 ||
             strcmp("listen_disabled_num", key) == 0 ||
             strcmp("conn_yields", key) == 0))
  {
    WATCHPOINT_STRING(key);
    return MEMCACHED_UNKNOWN_STAT_KEY;
  }

  return MEMCACHED_SUCCESS;
}

char *memcached_stat_get_value(memcached_st *ptr, memcached_stat_st *memc_stat, 
                               const char *key, memcached_return_t *error)
{
  char buffer[SMALL_STRING_LEN];
  int length;
  char *ret;

  *error= MEMCACHED_SUCCESS;

  if (!memcmp("pid", key, strlen("pid")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->pid);
  else if (!memcmp("uptime", key, strlen("uptime")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->uptime);
  else if (!memcmp("time", key, strlen("time")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->time);
  else if (!memcmp("version", key, strlen("version")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%s", memc_stat->version);
  else if (!memcmp("pointer_size", key, strlen("pointer_size")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->pointer_size);
  else if (!memcmp("rusage_user", key, strlen("rusage_user")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u.%u", memc_stat->rusage_user_seconds, memc_stat->rusage_user_microseconds);
  else if (!memcmp("rusage_system", key, strlen("rusage_system")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u.%u", memc_stat->rusage_system_seconds, memc_stat->rusage_system_microseconds);
  else if (!memcmp("curr_items", key, strlen("curr_items")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->curr_items);
  else if (!memcmp("total_items", key, strlen("total_items")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->total_items);
  else if (!memcmp("curr_connections", key, strlen("curr_connections")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->curr_connections);
  else if (!memcmp("total_connections", key, strlen("total_connections")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->total_connections);
  else if (!memcmp("connection_structures", key, strlen("connection_structures")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->connection_structures);
  else if (!memcmp("cmd_get", key, strlen("cmd_get")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->cmd_get);
  else if (!memcmp("cmd_set", key, strlen("cmd_set")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->cmd_set);
  else if (!memcmp("get_hits", key, strlen("get_hits")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->get_hits);
  else if (!memcmp("get_misses", key, strlen("get_misses")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->get_misses);
  else if (!memcmp("evictions", key, strlen("evictions")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->evictions);
  else if (!memcmp("bytes_read", key, strlen("bytes_read")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->bytes_read);
  else if (!memcmp("bytes_written", key, strlen("bytes_written")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->bytes_written);
  else if (!memcmp("bytes", key, strlen("bytes")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->bytes);
  else if (!memcmp("limit_maxbytes", key, strlen("limit_maxbytes")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%llu", (unsigned long long)memc_stat->limit_maxbytes);
  else if (!memcmp("threads", key, strlen("threads")))
    length= snprintf(buffer, SMALL_STRING_LEN,"%u", memc_stat->threads);
  else
  {
    *error= MEMCACHED_NOTFOUND;
    return NULL;
  }

  ret= ptr->call_malloc(ptr, (size_t) (length + 1));
  memcpy(ret, buffer, (size_t) length);
  ret[length]= '\0';

  return ret;
}

static memcached_return_t binary_stats_fetch(memcached_st *ptr,
                                             memcached_stat_st *memc_stat,
                                             char *args,
                                             unsigned int server_key)
{
  memcached_return_t rc;

  char buffer[MEMCACHED_DEFAULT_COMMAND_SIZE];
  protocol_binary_request_stats request= {.bytes= {0}};
  request.message.header.request.magic= PROTOCOL_BINARY_REQ;
  request.message.header.request.opcode= PROTOCOL_BINARY_CMD_STAT;
  request.message.header.request.datatype= PROTOCOL_BINARY_RAW_BYTES;

  if (args != NULL) 
  {
    size_t len= strlen(args);

    rc= memcached_validate_key_length(len, true);
    unlikely (rc != MEMCACHED_SUCCESS)
      return rc;

    request.message.header.request.keylen= htons((uint16_t)len);
    request.message.header.request.bodylen= htonl((uint32_t) len);

    if ((memcached_do(&ptr->hosts[server_key], request.bytes, 
                      sizeof(request.bytes), 0) != MEMCACHED_SUCCESS) ||
        (memcached_io_write(&ptr->hosts[server_key], args, len, 1) == -1)) 
    {
      memcached_io_reset(&ptr->hosts[server_key]);
      return MEMCACHED_WRITE_FAILURE;
    }
  }
  else
  {
    if (memcached_do(&ptr->hosts[server_key], request.bytes, 
                     sizeof(request.bytes), 1) != MEMCACHED_SUCCESS) 
    {
      memcached_io_reset(&ptr->hosts[server_key]);
      return MEMCACHED_WRITE_FAILURE;
    }
  }

  memcached_server_response_decrement(&ptr->hosts[server_key]);  
  do 
  {
    rc= memcached_response(&ptr->hosts[server_key], buffer, 
                           sizeof(buffer), NULL);
    if (rc == MEMCACHED_END)
      break;

    unlikely (rc != MEMCACHED_SUCCESS) 
    {
      memcached_io_reset(&ptr->hosts[server_key]);
      return rc;
    }

    unlikely((set_data(memc_stat, buffer, buffer + strlen(buffer) + 1)) == MEMCACHED_UNKNOWN_STAT_KEY)
    {
      WATCHPOINT_ERROR(MEMCACHED_UNKNOWN_STAT_KEY);
      WATCHPOINT_ASSERT(0);
    }
  } while (1);

  /* shit... memcached_response will decrement the counter, so I need to
   ** reset it.. todo: look at this and try to find a better solution.
 */
  ptr->hosts[server_key].cursor_active= 0;

  return MEMCACHED_SUCCESS;
}

static memcached_return_t ascii_stats_fetch(memcached_st *ptr,
                                            memcached_stat_st *memc_stat,
                                            char *args,
                                            unsigned int server_key)
{
  memcached_return_t rc;
  char buffer[MEMCACHED_DEFAULT_COMMAND_SIZE];
  size_t send_length;

  if (args)
    send_length= (size_t) snprintf(buffer, MEMCACHED_DEFAULT_COMMAND_SIZE, 
                                   "stats %s\r\n", args);
  else
    send_length= (size_t) snprintf(buffer, MEMCACHED_DEFAULT_COMMAND_SIZE, 
                                   "stats\r\n");

  if (send_length >= MEMCACHED_DEFAULT_COMMAND_SIZE)
    return MEMCACHED_WRITE_FAILURE;

  rc= memcached_do(&ptr->hosts[server_key], buffer, send_length, 1);
  if (rc != MEMCACHED_SUCCESS)
    goto error;

  while (1)
  {
    rc= memcached_response(&ptr->hosts[server_key], buffer, MEMCACHED_DEFAULT_COMMAND_SIZE, NULL);

    if (rc == MEMCACHED_STAT)
    {
      char *string_ptr, *end_ptr;
      char *key, *value;

      string_ptr= buffer;
      string_ptr+= 5; /* Move past STAT */
      for (end_ptr= string_ptr; isgraph(*end_ptr); end_ptr++);
      key= string_ptr;
      key[(size_t)(end_ptr-string_ptr)]= 0;

      string_ptr= end_ptr + 1;
      for (end_ptr= string_ptr; !(isspace(*end_ptr)); end_ptr++);
      value= string_ptr;
      value[(size_t)(end_ptr-string_ptr)]= 0;
      string_ptr= end_ptr + 2;
      unlikely((set_data(memc_stat, key, value)) == MEMCACHED_UNKNOWN_STAT_KEY)
      {
        WATCHPOINT_ERROR(MEMCACHED_UNKNOWN_STAT_KEY);
        WATCHPOINT_ASSERT(0);
      }
    }
    else
      break;
  }

error:
  if (rc == MEMCACHED_END)
    return MEMCACHED_SUCCESS;
  else
    return rc;
}

memcached_stat_st *memcached_stat(memcached_st *ptr, char *args, memcached_return_t *error)
{
  unsigned int x;
  memcached_return_t rc;
  memcached_stat_st *stats;

  unlikely (ptr->flags.use_udp)
  {
    *error= MEMCACHED_NOT_SUPPORTED;
    return NULL;
  }

  stats= ptr->call_calloc(ptr, ptr->number_of_hosts, sizeof(memcached_stat_st));

  if (!stats)
  {
    *error= MEMCACHED_MEMORY_ALLOCATION_FAILURE;
    return NULL;
  }

  rc= MEMCACHED_SUCCESS;
  for (x= 0; x < ptr->number_of_hosts; x++)
  {
    memcached_return_t temp_return;

    if (ptr->flags.binary_protocol)
      temp_return= binary_stats_fetch(ptr, stats + x, args, x);
    else
      temp_return= ascii_stats_fetch(ptr, stats + x, args, x);

    if (temp_return != MEMCACHED_SUCCESS)
      rc= MEMCACHED_SOME_ERRORS;
  }

  *error= rc;
  return stats;
}

memcached_return_t memcached_stat_servername(memcached_stat_st *memc_stat, char *args, 
                                             const char *hostname, in_port_t port)
{
  memcached_return_t rc;
  memcached_st memc;
  memcached_st *memc_ptr;

  memc_ptr= memcached_create(&memc);
  WATCHPOINT_ASSERT(memc_ptr);

  memcached_server_add(&memc, hostname, port);

  if (memc.flags.binary_protocol)
    rc= binary_stats_fetch(&memc, memc_stat, args, 0);
  else
    rc= ascii_stats_fetch(&memc, memc_stat, args, 0);

  memcached_free(&memc);

  return rc;
}

/* 
  We make a copy of the keys since at some point in the not so distant future
  we will add support for "found" keys.
*/
char ** memcached_stat_get_keys(memcached_st *ptr, memcached_stat_st *memc_stat, 
                                memcached_return_t *error)
{
  (void) memc_stat;
  char **list;
  size_t length= sizeof(memcached_stat_keys);

  list= ptr->call_malloc(ptr, length);

  if (!list)
  {
    *error= MEMCACHED_MEMORY_ALLOCATION_FAILURE;
    return NULL;
  }

  memcpy(list, memcached_stat_keys, sizeof(memcached_stat_keys));

  *error= MEMCACHED_SUCCESS;

  return list;
}

void memcached_stat_free(memcached_st *ptr, memcached_stat_st *memc_stat)
{
  if (memc_stat == NULL)
  {
    WATCHPOINT_ASSERT(0); /* Be polite, but when debugging catch this as an error */
    return;
  }

  if (ptr)
    ptr->call_free(ptr, memc_stat);
  else
    free(memc_stat);
}
